File: src/utils/builtinGateway.ts
/**
 * Loom built-in models channel (Gateway-X via public HTTPS).
 * Endpoint is fixed; credentials come from invite activation only.
 */

export const BUILTIN_GATEWAY_BASE = 'https://gateway.tanyun.store/v1';
export const BUILTIN_GATEWAY_ORIGIN = 'https://gateway.tanyun.store';
export const BUILTIN_PROFILE_ID = 'builtin-gateway';
export const BUILTIN_PROFILE_NAME = 'Loom Built-in';
/** OpenAI-compatible provider id used for stream/chat (not a new Rust provider). */
export const BUILTIN_TRANSPORT_PROVIDER = 'openai' as const;

export const BUILTIN_STORAGE_FILE = 'builtin-gateway.json';

export type BuiltinProtocolId = 'builtin';

export interface BuiltinQuotas {
  qps: number;
  daily_requests: number;
  daily_tokens: number;
}

/** Live `GET /v1/quota` snapshot (limits + today usage + remaining). */
export interface BuiltinQuotaStatus {
  quotas: BuiltinQuotas;
  usage: {
    daily_requests: number;
    daily_tokens: number;
  };
  /** null = unlimited (gateway uses JSON null when limit is 0). */
  remaining: {
    daily_requests: number | null;
    daily_tokens: number | null;
  };
}

export interface BuiltinGatewayState {
  installId: string;
  apiKey: string | null;
  /** HMAC signing secret from activation; never written to openai profile. */
  clientSecret: string | null;
  clientId: string | null;
  activatedAt: string | null;
  lastQuotas: BuiltinQuotas | null;
}

export interface BuiltinActivateResponse {
  api_key: string;
  client_secret: string;
  endpoint?: string;
  client_id: string;
  quotas?: Partial<BuiltinQuotas> | null;
}

export function isBuiltinProtocol(
  protocol: string | null | undefined
): protocol is BuiltinProtocolId {
  return protocol === 'builtin';
}

/** Logical UI/persistence providers that stream via the OpenAI-compatible path. */
export function isOpenaiCompatibleLogicalProvider(provider: string | null | undefined): boolean {
  return provider === 'openai' || provider === 'builtin';
}

/**
 * Map a logical protocol (UI / conversation) to the Rust stream provider.
 * Built-in models use the existing openai transport + synthetic profile.
 */
export function toTransportProvider(
  provider: string | null | undefined
): 'openai' | 'anthropic' | 'ollama' {
  if (provider === 'anthropic') return 'anthropic';
  if (provider === 'ollama') return 'ollama';
  // openai + builtin (+ unknown fallback)
  return 'openai';
}

/**
 * Profile id used when loading AI config / streaming for a logical provider.
 * Built-in always pins the injected openai profile.
 */
export function toTransportProfileId(
  provider: string | null | undefined,
  profileId?: string | null
): string | undefined {
  if (isBuiltinProtocol(provider)) {
    return BUILTIN_PROFILE_ID;
  }
  const trimmed = profileId?.trim();
  return trimmed || undefined;
}

/**
 * Provider key used when reading `profiles` / `configs` from load_ai_config.
 * Built-in models are stored under openai as profile `builtin-gateway`.
 */
export function toConfigProviderKey(
  provider: string | null | undefined
): 'openai' | 'anthropic' | 'ollama' {
  return toTransportProvider(provider);
}

export function buildTransportInvokeArgs(
  logicalProvider: string | null | undefined,
  model: string,
  profileId?: string | null
): {
  provider: 'openai' | 'anthropic' | 'ollama';
  model: string;
  profileId?: string;
} {
  return {
    provider: toTransportProvider(logicalProvider),
    model,
    profileId: toTransportProfileId(logicalProvider, profileId),
  };
}

export function keyPrefix(apiKey: string | null | undefined, keep = 12): string {
  if (!apiKey) return '';
  if (apiKey.length <= keep) return `${apiKey.slice(0, 4)}…`;
  return `${apiKey.slice(0, keep)}…`;
}

export function createInstallId(): string {
  if (typeof crypto !== 'undefined' && typeof crypto.randomUUID === 'function') {
    return crypto.randomUUID();
  }
  return `inst-${Date.now().toString(36)}-${Math.random().toString(36).slice(2, 12)}`;
}

export function normalizeActivateResponse(raw: unknown): BuiltinActivateResponse {
  if (!raw || typeof raw !== 'object') {
    throw new Error('Invalid activation response');
  }
  const obj = raw as Record<string, unknown>;
  const apiKey = typeof obj.api_key === 'string' ? obj.api_key : '';
  const clientSecret =
    typeof obj.client_secret === 'string'
      ? obj.client_secret
      : typeof obj.clientSecret === 'string'
        ? obj.clientSecret
        : '';
  const clientId = typeof obj.client_id === 'string' ? obj.client_id : '';
  if (!apiKey.trim() || !clientSecret.trim() || !clientId.trim()) {
    throw new Error('Activation response missing api_key, client_secret, or client_id');
  }
  let quotas: Partial<BuiltinQuotas> | null = null;
  if (obj.quotas && typeof obj.quotas === 'object') {
    const q = obj.quotas as Record<string, unknown>;
    quotas = {
      qps: typeof q.qps === 'number' ? q.qps : 0,
      daily_requests: typeof q.daily_requests === 'number' ? q.daily_requests : 0,
      daily_tokens: typeof q.daily_tokens === 'number' ? q.daily_tokens : 0,
    };
  }
  return {
    api_key: apiKey,
    client_secret: clientSecret,
    endpoint: typeof obj.endpoint === 'string' ? obj.endpoint : undefined,
    client_id: clientId,
    quotas,
  };
}

export function parseModelsListPayload(payload: unknown): string[] {
  if (!payload || typeof payload !== 'object') return [];
  const data = (payload as { data?: unknown }).data;
  if (!Array.isArray(data)) return [];
  const ids: string[] = [];
  for (const item of data) {
    if (item && typeof item === 'object' && typeof (item as { id?: unknown }).id === 'string') {
      const id = (item as { id: string }).id.trim();
      if (id) ids.push(id);
    }
  }
  return ids;
}

function readNonNegInt(value: unknown, fallback = 0): number {
  if (typeof value === 'number' && Number.isFinite(value) && value >= 0) {
    return Math.floor(value);
  }
  if (typeof value === 'string' && value.trim() && Number.isFinite(Number(value))) {
    const n = Number(value);
    return n >= 0 ? Math.floor(n) : fallback;
  }
  return fallback;
}

/** Parse Gateway-X `GET /v1/quota` JSON. */
export function parseQuotaStatusPayload(payload: unknown): BuiltinQuotaStatus | null {
  if (!payload || typeof payload !== 'object') return null;
  const obj = payload as Record<string, unknown>;
  const quotasObj =
    obj.quotas && typeof obj.quotas === 'object' ? (obj.quotas as Record<string, unknown>) : null;
  const usageObj =
    obj.usage && typeof obj.usage === 'object' ? (obj.usage as Record<string, unknown>) : null;
  const remainingObj =
    obj.remaining && typeof obj.remaining === 'object'
      ? (obj.remaining as Record<string, unknown>)
      : null;
  if (!quotasObj || !usageObj || !remainingObj) return null;

  const remainingOf = (v: unknown): number | null => {
    if (v === null || v === undefined) return null;
    return readNonNegInt(v, 0);
  };

  return {
    quotas: {
      qps: readNonNegInt(quotasObj.qps, 0),
      daily_requests: readNonNegInt(quotasObj.daily_requests, 0),
      daily_tokens: readNonNegInt(quotasObj.daily_tokens, 0),
    },
    usage: {
      daily_requests: readNonNegInt(usageObj.daily_requests, 0),
      daily_tokens: readNonNegInt(usageObj.daily_tokens, 0),
    },
    remaining: {
      daily_requests: remainingOf(remainingObj.daily_requests),
      daily_tokens: remainingOf(remainingObj.daily_tokens),
    },
  };
}

/** OpenAI-style gateway errors: `{ error: { message, type } }` or flat `{ message }` / `{ error: string }`. */
export function extractGatewayErrorMessage(payload: unknown, fallback = ''): string {
  if (!payload || typeof payload !== 'object') return fallback;
  const obj = payload as Record<string, unknown>;
  if (typeof obj.message === 'string' && obj.message.trim()) return obj.message.trim();
  const err = obj.error;
  if (typeof err === 'string' && err.trim()) return err.trim();
  if (err && typeof err === 'object') {
    const nested = err as Record<string, unknown>;
    if (typeof nested.message === 'string' && nested.message.trim()) {
      return nested.message.trim();
    }
  }
  return fallback;
}

/** Detect Gateway-X runtime auth failures in Rust/API error strings. */
export function isGatewayAuthErrorMessage(errorMsg: string): boolean {
  const text = errorMsg.trim();
  if (!text) return false;

  if (/client request signature|re-activate|重新激活|缺少 clientSecret|X-Gateway-/i.test(text)) {
    return true;
  }

  const jsonStart = text.indexOf('{');
  if (jsonStart >= 0) {
    try {
      const parsed = JSON.parse(text.slice(jsonStart)) as {
        error?: { type?: string; message?: string };
      };
      if (parsed.error?.type === 'auth_error') return true;
      const msg = parsed.error?.message ?? '';
      if (/unauthorized|re-activate|signature/i.test(msg)) return true;
    } catch {
      // not JSON — fall through
    }
  }

  return /\b401\b/.test(text) && /auth_error|unauthorized/i.test(text);
}

/** Map raw stream/API errors to a friendly built-in gateway message when appropriate. */
export function formatBuiltinGatewayStreamError(
  errorMsg: string,
  friendlyMessage: string,
  options?: { treatAsBuiltin?: boolean }
): string {
  const treatAsBuiltin = options?.treatAsBuiltin ?? false;
  if (treatAsBuiltin && /\b401\b/.test(errorMsg)) {
    return friendlyMessage;
  }
  if (isGatewayAuthErrorMessage(errorMsg)) {
    return friendlyMessage;
  }
  return errorMsg;
}

export function resolveBuiltinStreamError(
  errorMsg: string,
  friendlyMessage: string,
  options?: { treatAsBuiltin?: boolean }
): { message: string; unauthorized: boolean } {
  const message = formatBuiltinGatewayStreamError(errorMsg, friendlyMessage, options);
  return { message, unauthorized: message !== errorMsg };
}

export async function activateBuiltinGateway(
  inviteCode: string,
  installId: string,
  fetchImpl?: typeof fetch
): Promise<BuiltinActivateResponse> {
  const code = inviteCode.trim();
  if (!code) throw new Error('Invite code is required');
  if (!installId.trim()) throw new Error('install_id is required');

  // Prefer Tauri command (no CORS). Optional fetchImpl is for unit tests only.
  if (!fetchImpl) {
    try {
      const { invoke, isTauri } = await import('@tauri-apps/api/core');
      if (isTauri()) {
        const result = await invoke<Record<string, unknown>>('builtin_gateway_activate', {
          inviteCode: code,
          installId: installId.trim(),
        });
        return normalizeActivateResponse(result);
      }
    } catch (error) {
      const msg = error instanceof Error ? error.message : String(error);
      const err = new Error(msg) as Error & { status?: number };
      if (/401|unauthorized/i.test(msg)) err.status = 401;
      throw err;
    }
  }

  const doFetch = fetchImpl ?? fetch;
  const res = await doFetch(`${BUILTIN_GATEWAY_BASE}/activate`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json', Accept: 'application/json' },
    body: JSON.stringify({ invite_code: code, install_id: installId }),
  });

  const text = await res.text();
  let json: unknown;
  try {
    json = text ? JSON.parse(text) : null;
  } catch {
    json = null;
  }

  if (!res.ok) {
    const msg = extractGatewayErrorMessage(json) || text || `Activation failed (${res.status})`;
    const err = new Error(msg) as Error & { status?: number };
    err.status = res.status;
    throw err;
  }

  return normalizeActivateResponse(json);
}

export async function fetchBuiltinModels(
  apiKey: string,
  fetchImpl?: typeof fetch
): Promise<string[]> {
  if (!fetchImpl) {
    try {
      const { invoke, isTauri } = await import('@tauri-apps/api/core');
      if (isTauri()) {
        const result = await invoke<{ models: string[] }>('builtin_gateway_list_models', {
          apiKey,
        });
        return Array.isArray(result?.models) ? result.models : [];
      }
    } catch (error) {
      const msg = error instanceof Error ? error.message : String(error);
      const err = new Error(msg === 'UNAUTHORIZED' ? 'UNAUTHORIZED' : msg) as Error & {
        status?: number;
      };
      if (msg === 'UNAUTHORIZED' || /401|unauthorized/i.test(msg)) err.status = 401;
      throw err;
    }
  }

  const doFetch = fetchImpl ?? fetch;
  const res = await doFetch(`${BUILTIN_GATEWAY_BASE}/models`, {
    method: 'GET',
    headers: {
      Authorization: `Bearer ${apiKey}`,
      Accept: 'application/json',
    },
  });
  if (res.status === 401) {
    const err = new Error('UNAUTHORIZED') as Error & { status?: number };
    err.status = 401;
    throw err;
  }
  if (!res.ok) {
    const err = new Error(`Failed to list models (${res.status})`) as Error & { status?: number };
    err.status = res.status;
    throw err;
  }
  const json = await res.json();
  return parseModelsListPayload(json);
}

export async function checkBuiltinHealth(fetchImpl?: typeof fetch): Promise<boolean> {
  if (!fetchImpl) {
    try {
      const { invoke, isTauri } = await import('@tauri-apps/api/core');
      if (isTauri()) {
        const result = await invoke<{ ok: boolean }>('builtin_gateway_health');
        return Boolean(result?.ok);
      }
    } catch {
      return false;
    }
  }

  try {
    const doFetch = fetchImpl ?? fetch;
    const res = await doFetch(`${BUILTIN_GATEWAY_ORIGIN}/healthz`, { method: 'GET' });
    return res.ok;
  } catch {
    return false;
  }
}

export async function fetchBuiltinQuota(
  apiKey: string,
  fetchImpl?: typeof fetch
): Promise<BuiltinQuotaStatus> {
  if (!fetchImpl) {
    try {
      const { invoke, isTauri } = await import('@tauri-apps/api/core');
      if (isTauri()) {
        const result = await invoke<Record<string, unknown>>('builtin_gateway_get_quota', {
          apiKey,
        });
        const parsed = parseQuotaStatusPayload(result);
        if (!parsed) throw new Error('Invalid quota response');
        return parsed;
      }
    } catch (error) {
      const msg = error instanceof Error ? error.message : String(error);
      const err = new Error(msg === 'UNAUTHORIZED' ? 'UNAUTHORIZED' : msg) as Error & {
        status?: number;
      };
      if (msg === 'UNAUTHORIZED' || /401|unauthorized/i.test(msg)) err.status = 401;
      throw err;
    }
  }

  const doFetch = fetchImpl ?? fetch;
  const res = await doFetch(`${BUILTIN_GATEWAY_BASE}/quota`, {
    method: 'GET',
    headers: {
      Authorization: `Bearer ${apiKey}`,
      Accept: 'application/json',
    },
  });
  if (res.status === 401) {
    const err = new Error('UNAUTHORIZED') as Error & { status?: number };
    err.status = 401;
    throw err;
  }
  if (!res.ok) {
    const err = new Error(`Failed to load quota (${res.status})`) as Error & { status?: number };
    err.status = res.status;
    throw err;
  }
  const json = await res.json();
  const parsed = parseQuotaStatusPayload(json);
  if (!parsed) throw new Error('Invalid quota response');
  return parsed;
}

/** Shape used when merging into load_ai_config profiles.openai */
export function buildBuiltinProfileItem(apiKey: string, models: string[]) {
  return {
    id: BUILTIN_PROFILE_ID,
    name: BUILTIN_PROFILE_NAME,
    endpoint: BUILTIN_GATEWAY_BASE,
    apiKey,
    models: models.length > 0 ? models : [''],
    organizationId: undefined as string | undefined,
    supportsVision: false,
  };
}

/**
 * Ensure openai profiles contain the builtin gateway profile (in-memory + optional disk).
 * Does not remove user profiles. Marks activeId only when `makeActive` is true.
 */
export function mergeBuiltinProfileIntoAiConfig(
  config: Record<string, unknown>,
  apiKey: string,
  models: string[],
  options?: { makeActive?: boolean }
): Record<string, unknown> {
  const makeActive = options?.makeActive ?? false;
  const profiles = {
    ...((config.profiles as Record<string, unknown> | undefined) ?? {}),
  };
  const openai = {
    ...((profiles.openai as { activeId?: string; items?: unknown[] } | undefined) ?? {
      activeId: '',
      items: [],
    }),
  };
  const items = Array.isArray(openai.items) ? [...openai.items] : [];
  const item = buildBuiltinProfileItem(apiKey, models);
  const idx = items.findIndex(
    (it) => it && typeof it === 'object' && (it as { id?: string }).id === BUILTIN_PROFILE_ID
  );
  if (idx >= 0) {
    items[idx] = { ...(items[idx] as object), ...item };
  } else {
    items.push(item);
  }
  profiles.openai = {
    ...openai,
    items,
    activeId: makeActive
      ? BUILTIN_PROFILE_ID
      : openai.activeId || (items[0] as { id?: string })?.id,
  };
  const configs = {
    ...((config.configs as Record<string, unknown> | undefined) ?? {}),
  };
  // Keep configs.openai pointing at user active profile when not makeActive; when
  // makeActive, mirror builtin for legacy code paths that only read configs.
  if (makeActive) {
    configs.openai = {
      endpoint: BUILTIN_GATEWAY_BASE,
      apiKey,
      models: models.length > 0 ? models : [],
    };
  }
  return {
    ...config,
    profiles,
    configs,
  };
}

File: src-tauri/src/domain/ai/chat/builtin_gateway.rs
//! Built-in model channel (Gateway-X) HTTP helpers.
//! Requests run on the Rust side so the webview never hits CORS.

use super::gateway_sign::{apply_openai_gateway_auth, load_builtin_credentials};
use serde::Serialize;
use serde_json::Value;

const BUILTIN_GATEWAY_BASE: &str = "https://gateway.tanyun.store/v1";
const BUILTIN_GATEWAY_ORIGIN: &str = "https://gateway.tanyun.store";

fn build_client() -> Result<reqwest::Client, String> {
    reqwest::Client::builder()
        .connect_timeout(std::time::Duration::from_secs(15))
        .timeout(std::time::Duration::from_secs(30))
        .build()
        .map_err(|e| format!("创建HTTP客户端失败: {}", e))
}

#[derive(Debug, Serialize)]
pub struct BuiltinHealthResult {
    pub ok: bool,
}

#[derive(Debug, Serialize)]
pub struct BuiltinActivateResult {
    pub api_key: String,
    pub client_secret: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub endpoint: Option<String>,
    pub client_id: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub quotas: Option<Value>,
}

#[derive(Debug, Serialize)]
pub struct BuiltinModelsResult {
    pub models: Vec<String>,
}

#[derive(Debug, Serialize)]
pub struct BuiltinQuotaResult {
    pub quotas: Value,
    pub usage: Value,
    pub remaining: Value,
}

#[tauri::command]
pub async fn builtin_gateway_health() -> Result<BuiltinHealthResult, String> {
    let client = build_client()?;
    let url = format!("{}/healthz", BUILTIN_GATEWAY_ORIGIN);
    match client.get(&url).send().await {
        Ok(res) => Ok(BuiltinHealthResult {
            ok: res.status().is_success(),
        }),
        Err(_) => Ok(BuiltinHealthResult { ok: false }),
    }
}

#[tauri::command]
pub async fn builtin_gateway_activate(
    invite_code: String,
    install_id: String,
) -> Result<BuiltinActivateResult, String> {
    let code = invite_code.trim();
    let install = install_id.trim();
    if code.is_empty() {
        return Err("Invite code is required".to_string());
    }
    if install.is_empty() {
        return Err("install_id is required".to_string());
    }

    let client = build_client()?;
    let url = format!("{}/activate", BUILTIN_GATEWAY_BASE);
    let body = serde_json::json!({
        "invite_code": code,
        "install_id": install,
    });

    let res = client
        .post(&url)
        .header("Content-Type", "application/json")
        .header("Accept", "application/json")
        .json(&body)
        .send()
        .await
        .map_err(|e| format!("Activation request failed: {}", e))?;

    let status = res.status();
    let text = res
        .text()
        .await
        .map_err(|e| format!("Failed to read activation response: {}", e))?;

    let json: Value = if text.trim().is_empty() {
        Value::Null
    } else {
        serde_json::from_str(&text).unwrap_or(Value::Null)
    };

    if !status.is_success() {
        let msg = json
            .get("message")
            .and_then(|v| v.as_str())
            .or_else(|| json.get("error").and_then(|v| v.as_str()))
            .or_else(|| {
                json.get("error")
                    .and_then(|v| v.get("message"))
                    .and_then(|v| v.as_str())
            })
            .map(|s| s.trim().to_string())
            .filter(|s| !s.is_empty())
            .unwrap_or_else(|| {
                if text.trim().is_empty() {
                    format!("Activation failed ({})", status.as_u16())
                } else {
                    text
                }
            });
        return Err(msg);
    }

    let api_key = json
        .get("api_key")
        .and_then(|v| v.as_str())
        .unwrap_or("")
        .trim()
        .to_string();
    let client_secret = json
        .get("client_secret")
        .and_then(|v| v.as_str())
        .unwrap_or("")
        .trim()
        .to_string();
    let client_id = json
        .get("client_id")
        .and_then(|v| v.as_str())
        .unwrap_or("")
        .trim()
        .to_string();
    if api_key.is_empty() || client_secret.is_empty() || client_id.is_empty() {
        return Err(
            "Activation response missing api_key, client_secret, or client_id".to_string(),
        );
    }

    Ok(BuiltinActivateResult {
        api_key,
        client_secret,
        endpoint: json
            .get("endpoint")
            .and_then(|v| v.as_str())
            .map(|s| s.to_string()),
        client_id,
        quotas: json.get("quotas").cloned(),
    })
}

#[tauri::command]
pub async fn builtin_gateway_list_models(api_key: String) -> Result<BuiltinModelsResult, String> {
    let key = api_key.trim();
    if key.is_empty() {
        return Err("api_key is required".to_string());
    }

    let creds = load_builtin_credentials(None)?
        .ok_or_else(|| "内置网关需要重新激活（缺少 clientSecret）".to_string())?;

    let client = build_client()?;
    let url = format!("{}/models", BUILTIN_GATEWAY_BASE);
    let req = apply_openai_gateway_auth(
        client.get(&url),
        "GET",
        &url,
        &[],
        key,
        Some(&creds),
    )?;
    let res = req
        .header("Accept", "application/json")
        .send()
        .await
        .map_err(|e| format!("Failed to list models: {}", e))?;

    let status = res.status();
    if status.as_u16() == 401 {
        return Err("UNAUTHORIZED".to_string());
    }
    if !status.is_success() {
        return Err(format!("Failed to list models ({})", status.as_u16()));
    }

    let json: Value = res
        .json()
        .await
        .map_err(|e| format!("Failed to parse models response: {}", e))?;

    let mut models = Vec::new();
    if let Some(data) = json.get("data").and_then(|v| v.as_array()) {
        for item in data {
            if let Some(id) = item.get("id").and_then(|v| v.as_str()) {
                let trimmed = id.trim();
                if !trimmed.is_empty() {
                    models.push(trimmed.to_string());
                }
            }
        }
    }

    Ok(BuiltinModelsResult { models })
}

#[tauri::command]
pub async fn builtin_gateway_get_quota(api_key: String) -> Result<BuiltinQuotaResult, String> {
    let key = api_key.trim();
    if key.is_empty() {
        return Err("api_key is required".to_string());
    }

    let creds = load_builtin_credentials(None)?
        .ok_or_else(|| "内置网关需要重新激活（缺少 clientSecret）".to_string())?;

    let client = build_client()?;
    let url = format!("{}/quota", BUILTIN_GATEWAY_BASE);
    let req = apply_openai_gateway_auth(
        client.get(&url),
        "GET",
        &url,
        &[],
        key,
        Some(&creds),
    )?;
    let res = req
        .header("Accept", "application/json")
        .send()
        .await
        .map_err(|e| format!("Failed to load quota: {}", e))?;

    let status = res.status();
    if status.as_u16() == 401 {
        return Err("UNAUTHORIZED".to_string());
    }
    if !status.is_success() {
        return Err(format!("Failed to load quota ({})", status.as_u16()));
    }

    let json: Value = res
        .json()
        .await
        .map_err(|e| format!("Failed to parse quota response: {}", e))?;

    let quotas = json.get("quotas").cloned().unwrap_or(Value::Null);
    let usage = json.get("usage").cloned().unwrap_or(Value::Null);
    let remaining = json.get("remaining").cloned().unwrap_or(Value::Null);
    if !quotas.is_object() || !usage.is_object() || !remaining.is_object() {
        return Err("Invalid quota response".to_string());
    }

    Ok(BuiltinQuotaResult {
        quotas,
        usage,
        remaining,
    })
}
